presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+            for (Map.Entry<Symbol, FunctionCall> entry : scalarAggregation.getAggregations().entrySet()) {
+                FunctionCall call = entry.getValue();
+                QualifiedName count = QualifiedName.of("count");
+                Symbol symbol = entry.getKey();
+                if (call.getName().equals(count)) {
+                    aggregations.put(symbol, new FunctionCall(
+                            count,
+                            ImmutableList.of(nonNullableAggregationSourceSymbol.toSymbolReference())));
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
+        {
+            Optional<FilterNode> filter = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .findFirst();
+
+            if (!filter.isPresent()) {
+                return Optional.of(new RewrittenScalarAggregationSource(ImmutableList.of(), aggregationSource));
+            }
+
+            Expression predicate = filter.get().getPredicate();
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
```

Comment:
Since we're looking only at the first filter, the transformation will fail for queries that have multiple filters (say, separated by projections). This may or may not be a problem, depending on where in the sequence of optimizers this one sits.